# JSON Data Wrangling Exercise

This notebook takes data from the World Bank in the form of a JSON file and reads it in Python. The purpose of this notebook is to demonstrate familiarity with packages dealing with JSON and data wrangling techniques in Python.

Specific questions for this dataset are:
    + Find the top 10 co9untries with most projects.
    + Find the top 10 major project themes.
    + Create a dataframe with the missing names in the 'mjtheme_namecode' column filled in.

In [1]:
# Import pandas and JSON packages
import pandas as pd
import json
from pandas.io.json import json_normalize

In [2]:
# Assign fiel path to JSON file to variable to reduce typing redundance
json_filepath = 'data/world_bank_projects.json'

### **1. Find top 10 countries with most projects**

In [3]:
# Load JSON data as pandas dataframe
json_df = pd.read_json(json_filepath)
json_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [4]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

***
Now that the JSON file is in the form of a Panda DataFrame and there doesn't seem to be any missing values in the country columns, we can apply the 'value_count' method to get the top 10 countries with the most projects.
***

In [5]:
top_countries = json_df['countryname'].value_counts()
top_countries.head(10)

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Nepal                              12
Kingdom of Morocco                 12
People's Republic of Bangladesh    12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

### **2. Find the top 10 major project themes (using column 'mjtheme_namecode')**

In [6]:
# Read in JSON file as string
json_string = json.load((open(json_filepath)))

In [7]:
# Extract 'mjtheme_namecode' column using json_normalize
theme_namecode = json_normalize(json_string, 'mjtheme_namecode')
theme_namecode.head(10)

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

In [8]:
theme_namecode.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1499 entries, 0 to 1498
Data columns (total 2 columns):
code    1499 non-null object
name    1499 non-null object
dtypes: object(2)
memory usage: 23.5+ KB


***
**Note:** There's some missing values in the 'name' column, but that'll get cleaned up in the next exercise below. But there is a need to find out if the missing values are empty or blank spaces.
***

In [9]:
# What are the empty entries in the 'name' column?
theme_namecode['name'][1]

''

***
It seems that all blank entries in the 'name' column are just empty strings which will need to be handled. How many emtry strings are in the 'name' column? That'll determine if we can disregard those entries.
***

In [10]:
# How many emtry strings are there in the 'name' column?
(theme_namecode['name'] == '').sum()

122

***
There seems to be 122 out of 1499 entries that have a missing 'name' entry which accounts for ~8% of the data. For purposes of this exercise, I'll disregard the entries with empty strings and handle it in the next exercise.
***

In [11]:
# Filter out the rows with empty strings and group by the different themes
top_projects = theme_namecode[theme_namecode['name'] != ''].groupby(['code','name'])['name'].count().sort_values(ascending=False)
top_projects.head(10)

code  name                                        
11    Environment and natural resources management    223
10    Rural development                               202
8     Human development                               197
2     Public sector governance                        184
6     Social protection and risk management           158
4     Financial and private sector development        130
7     Social dev/gender/inclusion                     119
5     Trade and integration                            72
9     Urban development                                47
1     Economic management                              33
Name: name, dtype: int64

### **3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.**

In [12]:
# Use top_projects DataFrame to create dictionary with code# and name as key-value pairs
name_code = {(key):(value) for key, value in top_projects.sort_index().keys()}
name_code

{'1': 'Economic management',
 '10': 'Rural development',
 '11': 'Environment and natural resources management',
 '2': 'Public sector governance',
 '3': 'Rule of law',
 '4': 'Financial and private sector development',
 '5': 'Trade and integration',
 '6': 'Social protection and risk management',
 '7': 'Social dev/gender/inclusion',
 '8': 'Human development',
 '9': 'Urban development'}

In [13]:
# Replace values in 'name' based on their 'code' value by using the dictionary 'name_code' of names for each code
theme_namecode['name'] = [name_code[entry] for entry in theme_namecode.code.values]
theme_namecode.head(10)

code                                          name
0    8                             Human development
1   11  Environment and natural resources management
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

***
Recall that Index 1 in the DataFrame 'theme_namecode' from Exercise 2 was missing a name